### PR TITLE
postgresql 12.0

### PR DESCRIPTION
**Pull request by marksiemers**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

UPDATE: Looks like I had an issue with my local OS X SDK install. Followed the advice here: https://github.com/petere/homebrew-postgresql/issues/44#issuecomment-451487767
and now things are working.

First time trying to contribute by bumping to the latest postgresql.

Things were going pretty well, I made the changes so the audit would pass, ~~then building from source did not work:~~

### CODE REVIEW

**Comment by marksiemers** [comment]
> * New formulae should not have a `bottle do` block

What is the new preferred method?

I see this in the readme: https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#updating-formulae

**Comment by davidfetter** [comment]
I'd guess from the error message, leave out the `bottle do` block.

File: Formula/postgresql@11.rb [hunk 1]
```diff
@@ -0,0 +1,112 @@
+class PostgresqlAT11 < Formula
+  desc "Object-relational database system"
+  homepage "https://www.postgresql.org/"
+  url "https://ftp.postgresql.org/pub/source/v11.5/postgresql-11.5.tar.bz2"
+  sha256 "7fdf23060bfc715144cbf2696cf05b0fa284ad3eb21f0c378591c6bca99ad180"
+  revision 1
+  head "https://github.com/postgres/postgres.git"
+
+  bottle do
```
Comment (chenrui333) [inline Formula/postgresql@11.rb]: Just remove `bottle` block.

File: Formula/postgresql@11.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
+class PostgresqlAT11 < Formula
+  desc "Object-relational database system"
+  homepage "https://www.postgresql.org/"
+  url "https://ftp.postgresql.org/pub/source/v11.5/postgresql-11.5.tar.bz2"
+  sha256 "7fdf23060bfc715144cbf2696cf05b0fa284ad3eb21f0c378591c6bca99ad180"
+  revision 1
```
Comment (chenrui333) [inline Formula/postgresql@11.rb]: No need to have `revision` line

**Review by chenrui333 — CHANGES_REQUESTED** [review]

**Comment by chenrui333** [comment]
I think after the above two changes, it should be good.

Thanks @marksiemers !!

**Comment by marksiemers** [comment]
> I think after the above two changes, it should be good.
> 
> Thanks @marksiemers !!

Thank you!

I'm glad the changes are easy (hopefully). I have a couple work meetings this morning then I plan to update the PR later today.

**Comment by chenrui333** [comment]
Sounds good!

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
Thanks @marksiemers !!

**Comment by marksiemers** [comment]
@chenrui333 - Anything I can do to help with the failing builds? Are those failures expected?

**Comment by chenrui333** [comment]
a little bit busy at the moment, will take a look at it later.

From the build log, seems there are possible conflicts which might require revision bump:

```
14:49:15 Possible conflicting files are:
14:49:15 /usr/local/bin/clusterdb -> /usr/local/Cellar/postgresql@11/11.5/bin/clusterdb
14:49:15 /usr/local/bin/createdb -> /usr/local/Cellar/postgresql@11/11.5/bin/createdb
14:49:15 /usr/local/bin/createuser -> /usr/local/Cellar/postgresql@11/11.5/bin/createuser
14:49:15 /usr/local/bin/dropdb -> /usr/local/Cellar/postgresql@11/11.5/bin/dropdb
14:49:15 /usr/local/bin/dropuser -> /usr/local/Cellar/postgresql@11/11.5/bin/dropuser
14:49:15 /usr/local/bin/ecpg -> /usr/local/Cellar/postgresql@11/11.5/bin/ecpg
14:49:15 /usr/local/bin/initdb -> /usr/local/Cellar/postgresql@11/11.5/bin/initdb
14:49:15 /usr/local/bin/oid2name -> /usr/local/Cellar/postgresql@11/11.5/bin/oid2name
14:49:15 /usr/local/bin/pg_archivecleanup -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_archivecleanup
14:49:15 /usr/local/bin/pg_basebackup -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_basebackup
14:49:15 /usr/local/bin/pg_config -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_config
14:49:15 /usr/local/bin/pg_controldata -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_controldata
14:49:15 /usr/local/bin/pg_ctl -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_ctl
14:49:15 /usr/local/bin/pg_dump -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_dump
14:49:15 /usr/local/bin/pg_dumpall -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_dumpall
14:49:15 /usr/local/bin/pg_isready -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_isready
14:49:15 /usr/local/bin/pg_receivewal -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_receivewal
14:49:15 /usr/local/bin/pg_recvlogical -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_recvlogical
14:49:15 /usr/local/bin/pg_resetwal -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_resetwal
14:49:15 /usr/local/bin/pg_restore -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_restore
14:49:15 /usr/local/bin/pg_rewind -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_rewind
14:49:15 /usr/local/bin/pg_standby -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_standby
14:49:15 /usr/local/bin/pg_test_fsync -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_test_fsync
14:49:15 /usr/local/bin/pg_test_timing -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_test_timing
14:49:15 /usr/local/bin/pg_upgrade -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_upgrade
14:49:15 /usr/local/bin/pg_waldump -> /usr/local/Cellar/postgresql@11/11.5/bin/pg_waldump
14:49:15 /usr/local/bin/pgbench -> /usr/local/Cellar/postgresql@11/11.5/bin/pgbench
14:49:15 /usr/local/bin/postgres -> /usr/local/Cellar/postgresql@11/11.5/bin/postgres
14:49:15 /usr/local/bin/postmaster -> /usr/local/Cellar/postgresql@11/11.5/bin/postmaster
14:49:15 /usr/local/bin/psql -> /usr/local/Cellar/postgresql@11/11.5/bin/psql
14:49:15 /usr/local/bin/reindexdb -> /usr/local/Cellar/postgresql@11/11.5/bin/reindexdb
14:49:15 /usr/local/bin/vacuumdb -> /usr/local/Cellar/postgresql@11/11.5/bin/vacuumdb
14:49:15 /usr/local/bin/vacuumlo -> /usr/local/Cellar/postgresql@11/11.5/bin/vacuumlo
```

File: Formula/postgresql@11.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
+class PostgresqlAT11 < Formula
+  desc "Object-relational database system"
+  homepage "https://www.postgresql.org/"
+  url "https://ftp.postgresql.org/pub/source/v11.5/postgresql-11.5.tar.bz2"
+  sha256 "7fdf23060bfc715144cbf2696cf05b0fa284ad3eb21f0c378591c6bca99ad180"
+  head "https://github.com/postgres/postgres.git"
+
```
Comment (javian) [inline Formula/postgresql@11.rb]: You need to add `keg_only :versioned_formula` like its done in `postgresql@10`

File: Formula/postgresql@11.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    end
+  end
+
+  def caveats; <<~EOS
+    To migrate existing data from a previous major version of PostgreSQL run:
+      brew postgresql-upgrade-database
+  EOS
+  end
+
+  plist_options :manual => "pg_ctl -D #{HOMEBREW_PREFIX}/var/postgres start"
```
Comment (javian) [inline Formula/postgresql@11.rb]: This also needs to be updated for the versioned formula. If you `diff` the `postgres` formula and the `postgres@10` there might be more clues as to what changed between the last time we bumped the major version of postgres.

Comment (marksiemers) [reply Formula/postgresql@11.rb]: I'll continue to work through the issues.

I'm now seeing the commands that are running in CI that show failing tests, it looks like for at least some of them I can reproduce locally and shorten the feedback loop.

When I did the audit and build from source (when originally opening the PR) it was all for postgres 12 - I didn't run any real tests for postgresql@11 (at that time)

**Comment by marksiemers** [comment]
For these missing libraries:
```
Missing libraries:
  /usr/local/lib/libecpg.6.dylib
  /usr/local/lib/libpgtypes.3.dylib
  /usr/local/lib/libpq.5.dylib
```
Should `libpq` be listed as a dependency for postgresql@11? Or does it need to be available as a native library on the system (without using brew)?

Even with that, it might needs some compiler flags:
```
libpq is keg-only, which means it was not symlinked into /usr/local,
because conflicts with postgres formula.

If you need to have libpq first in your PATH run:
  echo 'export PATH="/usr/local/opt/libpq/bin:$PATH"' >> ~/.bash_profile

For compilers to find libpq you may need to set:
  export LDFLAGS="-L/usr/local/opt/libpq/lib"
  export CPPFLAGS="-I/usr/local/opt/libpq/include"

For pkg-config to find libpq you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/libpq/lib/pkgconfig"
```

**Comment by marksiemers** [comment]
This may be dependent on citus 9.0.0 and some others libs upgrading to support pg 12.

https://github.com/citusdata/citus/pull/3081

**Comment by hanefi** [comment]
I did not test building citus from source using the HomeBrew PostgreSQL. AFAIK, libpq is not needed, and it suffices to have `pg_config` in the `$PATH` (or the path of `pg_config`in `$PG_CONFIG`).

I will try to test it later today, and inform you on this thread.

**Comment by hanefi** [comment]
I confirmed my theory. It is sufficient to have `pg_config`, and libpq is not necessary to compile citus.

Even if it were, I think it is not the responsiblity of the Postgres package to have that dependency just because a postgres extension may need it.

**Comment by marksiemers** [comment]
This is the error that I'm struggling to get past:
```
Testing citus
==> /usr/local/opt/postgresql/bin/initdb /private/tmp/citus-test-20191008-6167-1ksbyh1/test
2019-10-08 09:08:33.697 PDT [6193] FATAL:  could not load library "/usr/local/lib/postgresql/citus.so": dlopen(/usr/local/lib/postgresql/citus.so, 10): Symbol not found: _AllocSetContextCreateExtended
	  Referenced from: /usr/local/lib/postgresql/citus.so
	  Expected in: /usr/local/opt/postgresql/bin/postgres
	 in /usr/local/lib/postgresql/citus.so
2019-10-08 09:08:33.697 PDT [6193] LOG:  database system is shut down
==> /usr/local/opt/postgresql/bin/createdb -p 55561 test
Last 15 lines from /Users/mark/Library/Logs/Homebrew/citus/test.02.createdb:
2019-10-08 09:08:35 -0700

/usr/local/opt/postgresql/bin/createdb
-p
55561
test

createdb: error: could not connect to database template1: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/tmp/.s.PGSQL.55561"?
```

**Comment by chenrui333** [comment]
@marksiemers just added a `help wanted` tag so that we can collaborate on this.

**Comment by marksiemers** [comment]
Citus 9.0.0 has been merged to master but no release yet.

There is a chance that bumping the formula for citus will resolve one of the errors in CI.

If that is the case, will a Formula be needed for `citus@8` to be compatible with some of the older postgres versions?

Or is citus 9.0.0 compatible with postgres 10 and 11 as well? @hanefi - can you comment on citus 9.x backwards compatibility?

**Comment by MatthiasWinkelmann** [comment]
Postgres 10 support was just removed in https://github.com/citusdata/citus/commit/74cb168205efbfdf03b8ba71b1de92154855a7f1. Note that this happened immediately *after* the 9.0 release, which should, therefore, support PG 10 and 11.

**Comment by marksiemers** [comment]
I haven't been able to spend much time with this recently.

I'm on Catalina now, and trying a fresh postgresql (12) with brew - install gives me problems with perl headers. It seems like this happens regularly where XCode moves the location of `perl.h` - so if anyone has a good idea of how to handle this in the postgresql Formula, it would be great to incorporate.

I can do the install `--without-perl` but that seems to cause downstream issues.

**Comment by xales** [comment]
@marksiemers Experimenting with this now - got it to work clean on a Catalina install - builds fine and `brew test` passes, anyway. Ran into the Perl header issue, I ended up adding brewed `perl` as a dependency to work around it (manually installing headers from the .pkg works too). Unsure what the preferred thing to do here is.

**Comment by tmepple** [comment]
At this point 12.1 was released yesterday.  Should the PR be updated or will that require a new PR?

**Comment by marksiemers** [comment]
@xales - Do you think a re-build of the existing code should work (assuming catalina has been updated in the CI environment)?

Or does the dependency on brew's perl need to be put into place?

@tmepple - I don't see any precedent for brew to support minor versions, just the latest major version. The only thing to look out for about being quick to upgrade minor versions is that other packages might not be ready. 

For example `postgis` didn't have a stable release that was postgresql 12 compatible until several weeks after the official PG 12 release.

**Comment by SMillerDev** [comment]
Yeah, this pr should be updated to 12.1

**Comment by lhaig** [comment]
How do we use this now? I just need pg_config to be there :-)

It would be wonderful if this was approved and usable soon. Thank you

**Comment by SMillerDev** [comment]
I'm ready to merge this, as soon as the build isn't failing.

**Comment by Bo98** [comment]
I'll be willing to help out here if you've hit a wall.

**Comment by Bo98** [comment]
Let's start again: #46893.

**Pull request closed without merging**